File: src/legal_format_engine/ml/rule_hierarchy.py
"""Rule hierarchy: court rules ALWAYS win over ML-learned patterns.

This is the gatekeeper between the ML system and the formatting engine.
It enforces a strict hierarchy:

    1. COURT RULES (YAML rulesets) — mandatory, never overridden
    2. ML-LEARNED PATTERNS — consensus from uploaded documents
    3. STYLE PROFILES (author/firm) — discretionary preferences
    4. ENGINE DEFAULTS — fallback for anything not specified

When a court rule specifies a value (e.g., "font_size_pt: 12"), no amount
of ML evidence can override it. If 100 uploaded briefs use 13pt font but
the rule says 12pt, the output is 12pt. Period.

When the rule is SILENT on a field (e.g., first-line indent, block quote
formatting), the ML and style profiles fill in with learned values.

This module:
- Identifies which formatting fields are rule-governed vs. discretionary
- Filters ML recommendations to only apply to discretionary areas
- Merges rule + ML + style profile into a final formatting spec
- Provides clear provenance for every formatting decision
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from legal_format_engine.ml.learner import LearnedFormat
from legal_format_engine.ml.styles import StyleProfile
from legal_format_engine.rules.schema import Ruleset


@dataclass
class FormattingDecision:
    """A single formatting decision with full provenance."""
    field: str  # e.g., "page_format.font_size_pt"
    value: Any  # the decided value
    source: str  # "rule", "ml_learned", "style_profile", "default"
    confidence: float = 1.0  # 1.0 for rules, varies for ML/style
    explanation: str = ""


@dataclass
class ResolvedFormat:
    """Complete formatting specification with provenance for every decision.

    Every formatting field has a value and a clear source explaining
    where it came from. This is what the engine uses to format documents.
    """
    jurisdiction: str
    court_level: str
    document_type: str

    decisions: list[FormattingDecision] = field(default_factory=list)

    # Convenience: count by source
    @property
    def rule_count(self) -> int:
        return sum(1 for d in self.decisions if d.source == "rule")

    @property
    def ml_count(self) -> int:
        return sum(1 for d in self.decisions if d.source == "ml_learned")

    @property
    def style_count(self) -> int:
        return sum(1 for d in self.decisions if d.source == "style_profile")

    @property
    def default_count(self) -> int:
        return sum(1 for d in self.decisions if d.source == "default")

    def get(self, field_name: str) -> FormattingDecision | None:
        for d in self.decisions:
            if d.field == field_name:
                return d
        return None

    def to_dict(self) -> dict:
        return {
            "jurisdiction": self.jurisdiction,
            "court_level": self.court_level,
            "document_type": self.document_type,
            "decisions": [
                {
                    "field": d.field,
                    "value": d.value,
                    "source": d.source,
                    "confidence": d.confidence,
                    "explanation": d.explanation,
                }
                for d in self.decisions
            ],
            "summary": {
                "total": len(self.decisions),
                "from_rules": self.rule_count,
                "from_ml": self.ml_count,
                "from_style": self.style_count,
                "from_defaults": self.default_count,
            },
        }


# Fields that are ALWAYS governed by court rules when present in the YAML.
# If the ruleset specifies these, the value is law. No ML override.
RULE_GOVERNED_FIELDS = {
    # Page format — courts specify these
    "font_name",
    "font_size_pt",
    "line_spacing",
    "margin_top_inches",
    "margin_bottom_inches",
    "margin_left_inches",
    "margin_right_inches",
    "page_width_inches",
    "page_height_inches",

    # Heading rules — courts often specify level formatting
    "heading_case_style",
    "heading_alignment",
    "heading_numbering",
    "heading_bold",

    # Required sections — courts mandate what must be included
    "required_sections",
    "section_order",

    # Caption formatting — courts specify this
    "caption_court_line_style",
    "caption_case_number_prefix",
    "caption_party_separator",
    "caption_party_name_style",

    # Certifications — specific legal language required
    "certifications",
}


def identify_rule_governed_fields(ruleset: Ruleset) -> set[str]:
    """Identify which fields are explicitly specified by the court ruleset.

    A field is rule-governed if the ruleset provides a non-default value
    for it. Fields where the ruleset is silent are discretionary.
    """
    governed = set()

    pf = ruleset.page_format

    # Check each page format field against defaults
    defaults = {
        "font_name": "Times New Roman",
        "font_size_pt": 12,
        "line_spacing": 2.0,
        "margin_top_inches": 1.0,
        "margin_bottom_inches": 1.0,
        "margin_left_inches": 1.0,
        "margin_right_inches": 1.0,
    }

    # If the ruleset specifies ANY value for these (even the default),
    # treat them as rule-governed. The YAML files we wrote explicitly
    # set these values, so they're intentional.
    governed.add("font_name")
    governed.add("font_size_pt")
    governed.add("line_spacing")
    governed.add("margin_top_inches")
    governed.add("margin_bottom_inches")
    governed.add("margin_left_inches")
    governed.add("margin_right_inches")

    # Heading rules: governed at the level they're defined
    for hr in ruleset.heading_rules:
        governed.add(f"heading_{hr.level}_case_style")
        governed.add(f"heading_{hr.level}_alignment")
        governed.add(f"heading_{hr.level}_bold")
        if hr.numbering is not None:
            governed.add(f"heading_{hr.level}_numbering")

    # Required sections
    if ruleset.required_sections:
        governed.add("required_sections")
        governed.add("section_order")

    # Caption
    governed.add("caption_style")

    # Certifications
    if ruleset.certifications:
        governed.add("certifications")

    return governed


def resolve_format(
    ruleset: Ruleset,
    learned: LearnedFormat | None = None,
    style: StyleProfile | None = None,
) -> ResolvedFormat:
    """Resolve final formatting by applying the hierarchy:

    Rule > ML Learned > Style Profile > Default

    Every decision is tracked with provenance.
    """
    resolved = ResolvedFormat(
        jurisdiction=ruleset.jurisdiction,
        court_level=ruleset.court_level,
        document_type=ruleset.document_type,
    )

    governed = identify_rule_governed_fields(ruleset)
    pf = ruleset.page_format

    # ------- TYPOGRAPHY -------

    # Font name: ALWAYS from rule
    resolved.decisions.append(FormattingDecision(
        field="font_name",
        value=pf.font_name,
        source="rule",
        confidence=1.0,
        explanation=f"Court rule requires {pf.font_name}",
    ))

    # Font size: ALWAYS from rule
    resolved.decisions.append(FormattingDecision(
        field="font_size_pt",
        value=pf.font_size_pt,
        source="rule",
        confidence=1.0,
        explanation=f"Court rule requires {pf.font_size_pt}pt",
    ))

    # Line spacing: ALWAYS from rule
    resolved.decisions.append(FormattingDecision(
        field="line_spacing",
        value=pf.line_spacing,
        source="rule",
        confidence=1.0,
        explanation=f"Court rule requires {pf.line_spacing}x spacing",
    ))

    # ------- MARGINS -------

    for margin_field in ("margin_top_inches", "margin_bottom_inches",
                         "margin_left_inches", "margin_right_inches"):
        val = getattr(pf, margin_field)
        resolved.decisions.append(FormattingDecision(
            field=margin_field,
            value=val,
            source="rule",
            confidence=1.0,
            explanation=f"Court rule requires {val}\" {margin_field.split('_')[1]} margin",
        ))

    # ------- HEADINGS -------

    for hr in ruleset.heading_rules:
        lvl = hr.level

        resolved.decisions.append(FormattingDecision(
            field=f"heading_{lvl}_case_style",
            value=hr.case_style,
            source="rule",
            confidence=1.0,
            explanation=f"Court rule: level {lvl} headings use {hr.case_style} case",
        ))

        resolved.decisions.append(FormattingDecision(
            field=f"heading_{lvl}_alignment",
            value=hr.alignment,
            source="rule",
            confidence=1.0,
            explanation=f"Court rule: level {lvl} headings are {hr.alignment}-aligned",
        ))

        resolved.decisions.append(FormattingDecision(
            field=f"heading_{lvl}_bold",
            value=hr.bold,
            source="rule",
            confidence=1.0,
            explanation=f"Court rule: level {lvl} headings {'bold' if hr.bold else 'not bold'}",
        ))

        if hr.numbering is not None:
            resolved.decisions.append(FormattingDecision(
                field=f"heading_{lvl}_numbering",
                value=hr.numbering,
                source="rule",
                confidence=1.0,
                explanation=f"Court rule: level {lvl} numbering is {hr.numbering}",
            ))

    # ------- DISCRETIONARY FIELDS (ML + Style fill the gaps) -------

    # Body paragraph first-line indent (rules almost never specify this)
    _resolve_discretionary(
        resolved, "body_first_line_indent",
        learned_attr="body_indent",
        learned_format=learned,
        style=style,
        default=0.5,
        unit="inches",
    )

    # Block quote indent
    _resolve_discretionary(
        resolved, "block_quote_indent",
        learned_attr="block_quote_indent",
        learned_format=learned,
        style=style,
        default=0.5,
        unit="inches",
    )

    # Heading spacing (discretionary)
    _resolve_discretionary(
        resolved, "heading_space_before_pt",
        learned_attr=None,
        learned_format=None,
        style=style,
        default=12.0,
        unit="pt",
    )
    _resolve_discretionary(
        resolved, "heading_space_after_pt",
        learned_attr=None,
        learned_format=None,
        style=style,
        default=6.0,
        unit="pt",
    )

    # Heading numbering for levels where rule is silent
    for level in (1, 2, 3, 4):
        numbering_field = f"heading_{level}_numbering"
        if numbering_field not in governed:
            _resolve_discretionary_heading_numbering(
                resolved, level, learned, style,
            )

    return resolved


def _resolve_discretionary(
    resolved: ResolvedFormat,
    field_name: str,
    learned_attr: str | None,
    learned_format: LearnedFormat | None,
    style: StyleProfile | None,
    default: Any,
    unit: str = "",
) -> None:
    """Resolve a discretionary field through the hierarchy: ML > Style > Default."""

    # Try ML learned value first
    if learned_format and learned_attr:
        learned_val = getattr(learned_format, learned_attr, None)
        if learned_val is not None and learned_val.confidence >= 0.3:
            resolved.decisions.append(FormattingDecision(
                field=field_name,
                value=learned_val.value,
                source="ml_learned",
                confidence=learned_val.confidence,
                explanation=(
                    f"Learned from {learned_val.sample_count} documents "
                    f"({learned_val.agreement:.0%} agreement)"
                ),
            ))
            return

    # Try style profile
    if style:
        pref = style.get(field_name)
        if pref is not None and pref.consistency >= 0.5:
            resolved.decisions.append(FormattingDecision(
                field=field_name,
                value=pref.value,
                source="style_profile",
                confidence=pref.consistency,
                explanation=(
                    f"Author/firm preference from {pref.sample_count} documents "
                    f"({pref.consistency:.0%} consistent)"
                ),
            ))
            return

    # Fall back to default
    unit_str = f"{unit} " if unit else ""
    resolved.decisions.append(FormattingDecision(
        field=field_name,
        value=default,
        source="default",
        confidence=0.5,
        explanation=f"Engine default: {default}{unit_str}(rule silent, no ML data)",
    ))


def _resolve_discretionary_heading_numbering(
    resolved: ResolvedFormat,
    level: int,
    learned: LearnedFormat | None,
    style: StyleProfile | None,
) -> None:
    """Resolve heading numbering when the rule doesn't specify it."""
    field_name = f"heading_{level}_numbering"

    # Check ML learned heading styles
    if learned:
        for hs in learned.heading_styles:
            if hs.level == level and hs.numbering and hs.numbering.confidence >= 0.3:
                resolved.decisions.append(FormattingDecision(
                    field=field_name,
                    value=hs.numbering.value,
                    source="ml_learned",
                    confidence=hs.numbering.confidence,
                    explanation=(
                        f"Learned from documents: level {level} headings "
                        f"typically use {hs.numbering.value} numbering"
                    ),
                ))
                return

    # No learned data — use conventional defaults
    level_defaults = {
        1: None,
        2: "roman",
        3: "alpha_upper",
        4: "arabic",
    }
    resolved.decisions.append(FormattingDecision(
        field=field_name,
        value=level_defaults.get(level),
        source="default",
        confidence=0.5,
        explanation=f"Engine default numbering for level {level}",
    ))

File: src/LegalEngine.Rules/RuleHierarchy.fs
namespace LegalEngine.Rules

/// Enforces the strict formatting rule hierarchy:
///
///   1. Court Rules (YAML rulesets)  — MANDATORY, never overridden
///   2. ML Learned Patterns (JSON)   — fills discretionary gaps only
///   3. Style Profiles (per-firm)    — lowest priority override
///   4. Engine Defaults              — fallback for everything else
///
/// The court rules are the floor. ML can refine within them but never violate.
/// A formatting decision is "discretionary" if the court rule does not specify it.

open LegalEngine.Domain


/// Where a formatting decision came from.
[<RequireQualifiedAccess>]
type DecisionSource =
    | CourtRule        // Mandatory — from YAML ruleset
    | MlLearned        // Advisory — from ML pattern analysis
    | StyleProfile     // Preference — from firm/author style
    | Default          // Fallback — engine default

/// A single formatting decision with provenance.
type FormattingDecision = {
    Field: string
    Value: obj
    Source: DecisionSource
    Confidence: float
    Explanation: string
}

/// Fully resolved format with provenance for every decision.
type ResolvedFormat = {
    Jurisdiction: string
    CourtLevel: string
    DocumentType: string
    Decisions: FormattingDecision list
    PageFormat: PageFormat
    HeadingRules: HeadingRule list
}


/// Style profile for a firm or author (matches ML engine output).
type StyleProfile = {
    Identifier: string
    ProfileType: string  // "author" or "firm"
    FirstLineIndent: float option
    BlockQuoteIndent: float option
    SectionSpacing: float option
    Justify: bool option
    DocumentCount: int
    Confidence: float
}

module StyleProfile =
    let empty = {
        Identifier = ""
        ProfileType = "author"
        FirstLineIndent = None
        BlockQuoteIndent = None
        SectionSpacing = None
        Justify = None
        DocumentCount = 0
        Confidence = 0.0
    }


module RuleHierarchy =

    /// Resolve the final formatting by applying the 4-tier hierarchy.
    /// Court rules always win. ML fills gaps. Style profiles are lowest override. Defaults are fallback.
    let resolve
        (ruleset: Ruleset)
        (mlPatterns: MlPatternSet option)
        (styleProfile: StyleProfile option)
        : ResolvedFormat =

        let decisions = ResizeArray<FormattingDecision>()

        // Start with defaults
        let mutable pageFormat = PageFormat.defaults

        // ── Layer 4: Engine Defaults (already set) ──

        // ── Layer 3: Style Profile (overrides defaults for discretionary fields) ──
        match styleProfile with
        | Some style ->
            match style.FirstLineIndent with
            | Some indent ->
                pageFormat <- { pageFormat with FirstLineIndentInches = indent }
                decisions.Add {
                    Field = "first_line_indent_inches"
                    Value = indent
                    Source = DecisionSource.StyleProfile
                    Confidence = style.Confidence
                    Explanation = $"Style profile '{style.Identifier}' prefers {indent}\" indent"
                }
            | None -> ()
        | None -> ()

        // ── Layer 2: ML Learned Patterns (overrides style for discretionary fields) ──
        match mlPatterns with
        | Some ml ->
            // First-line indent (discretionary — ML can override)
            match ml.FirstLineIndent with
            | Some learned when learned.Confidence >= 0.5 ->
                pageFormat <- { pageFormat with FirstLineIndentInches = learned.Value }
                decisions.Add {
                    Field = "first_line_indent_inches"
                    Value = learned.Value
                    Source = DecisionSource.MlLearned
                    Confidence = learned.Confidence
                    Explanation = $"ML learned {learned.Value}\" indent from {learned.SampleCount} documents"
                }
            | _ -> ()

            // Block quote indent (discretionary)
            match ml.BlockQuoteIndent with
            | Some learned when learned.Confidence >= 0.5 ->
                decisions.Add {
                    Field = "block_quote_indent_inches"
                    Value = learned.Value
                    Source = DecisionSource.MlLearned
                    Confidence = learned.Confidence
                    Explanation = $"ML learned {learned.Value}\" block quote indent"
                }
            | _ -> ()
        | None -> ()

        // ── Layer 1: Court Rules (MANDATORY — overrides EVERYTHING) ──
        let courtFormat = ruleset.PageFormat

        // Font — court rule always wins
        pageFormat <- { pageFormat with FontName = courtFormat.FontName }
        decisions.Add {
            Field = "font_name"; Value = courtFormat.FontName
            Source = DecisionSource.CourtRule; Confidence = 1.0
            Explanation = "Court rule requires this font"
        }

        pageFormat <- { pageFormat with FontSizePt = courtFormat.FontSizePt }
        decisions.Add {
            Field = "font_size_pt"; Value = courtFormat.FontSizePt
            Source = DecisionSource.CourtRule; Confidence = 1.0
            Explanation = "Court rule requires this font size"
        }

        // Line spacing — court rule
        pageFormat <- { pageFormat with LineSpacing = courtFormat.LineSpacing }
        decisions.Add {
            Field = "line_spacing"; Value = courtFormat.LineSpacing
            Source = DecisionSource.CourtRule; Confidence = 1.0
            Explanation = "Court rule requires this line spacing"
        }

        // Margins — court rule
        pageFormat <- { pageFormat with
                            MarginTopInches = courtFormat.MarginTopInches
                            MarginBottomInches = courtFormat.MarginBottomInches
                            MarginLeftInches = courtFormat.MarginLeftInches
                            MarginRightInches = courtFormat.MarginRightInches }
        decisions.Add {
            Field = "margins"; Value = "court-mandated"
            Source = DecisionSource.CourtRule; Confidence = 1.0
            Explanation = $"Court rule: T={courtFormat.MarginTopInches}\" B={courtFormat.MarginBottomInches}\" L={courtFormat.MarginLeftInches}\" R={courtFormat.MarginRightInches}\""
        }

        // Page dimensions — court rule
        pageFormat <- { pageFormat with
                            PageWidthInches = courtFormat.PageWidthInches
                            PageHeightInches = courtFormat.PageHeightInches }

        // Body alignment — court rule
        pageFormat <- { pageFormat with BodyAlignment = courtFormat.BodyAlignment }

        // First-line indent: court rule wins ONLY if it explicitly sets a value
        // If the court rule uses the default (0.5), ML/style can override
        // This is the "discretionary gap" concept
        if courtFormat.FirstLineIndentInches <> PageFormat.defaults.FirstLineIndentInches then
            pageFormat <- { pageFormat with FirstLineIndentInches = courtFormat.FirstLineIndentInches }
            decisions.Add {
                Field = "first_line_indent_inches"
                Value = courtFormat.FirstLineIndentInches
                Source = DecisionSource.CourtRule
                Confidence = 1.0
                Explanation = "Court rule explicitly sets first-line indent"
            }

        // Heading rules: court rules always win
        let headingRules = ruleset.HeadingRules

        // ML heading styles are advisory — only produce INFO findings, never override
        match mlPatterns with
        | Some ml ->
            for learned in ml.HeadingStyles do
                let courtRule = headingRules |> List.tryFind (fun r -> r.Level = learned.Value.Level)
                match courtRule with
                | Some cr when cr.CaseStyle <> learned.Value.CaseStyle && learned.Confidence >= 0.7 ->
                    decisions.Add {
                        Field = $"heading_level_{learned.Value.Level}_case_style"
                        Value = learned.Value.CaseStyle
                        Source = DecisionSource.MlLearned
                        Confidence = learned.Confidence
                        Explanation = $"ML suggests '{learned.Value.CaseStyle}' but court rule requires '{cr.CaseStyle}' — court rule wins"
                    }
                | _ -> ()
        | None -> ()

        {
            Jurisdiction = ruleset.Jurisdiction
            CourtLevel = ruleset.CourtLevel
            DocumentType = ruleset.DocumentType
            Decisions = decisions |> Seq.toList
            PageFormat = pageFormat
            HeadingRules = headingRules
        }

    /// Check if a field is governed by court rules (mandatory) or discretionary.
    let isDiscretionary (field: string) : bool =
        let mandatoryFields = set [
            "font_name"; "font_size_pt"; "line_spacing"
            "margin_top_inches"; "margin_bottom_inches"
            "margin_left_inches"; "margin_right_inches"
            "page_width_inches"; "page_height_inches"
        ]
        not (mandatoryFields.Contains field)
